[GOV] proposed code of conduct update - mature sktime structure and decision making on resources

fkiraly opened the pull request:
Changes of substance compared to the original PR #3095:

* upon suggestion of @GuzalBulatova (below), clarified the scope of applicability, including a clarification that the general terms of license by which the software is distributed are unaffected. The scope is clarified to apply, primarily, to elected members of the community in their elected roles (while excluding loopholes like saying "I did this in my other role, actually").
* upon suggestion of @GuzalBulatova (slack), changed expectations on creating opportunities. In the new version, leaders are expected to create fair opportunities (without restriction), and members of the community are strongly encouraged (not expected ) to create fair opportunities as have been created for them.
* upon request of @aiwalter (slack), removed expectation that majority of benefits of an `sktime` commercial venture need to go to the community. What remains is the weaker requirement that it needs to go through transparent information channels and community decision making.
* upon request of @aiwalter (PR #3095), added clarification on the meaning of "resource"
* upon suggestion of @GuzalBulatova (below), removed "anxious collaborators" - this was not added by the original PR #3095 but @GuzalBulatova requested the change nonetheless. I'm fine with that.

Minor changes:

* reformulation on the resource allocation points in Section 2.2
* some reformulations/clarifications in "investigating CoC violations on resource allocation" appendix section

---

**ANY COMMENTS/DISCUSSION ON THE ORIGINAL PR PLEASE**
https://github.com/sktime/sktime/pull/3095

---

Original description below. No longer a revert, as it makes changes above #3095.

---

Reverts sktime/sktime#3673.

@aiwalter, I do not think this was a violation of our rules, but I agree there may be need for discussion since there were some changes last night and since this PR is now at the center of attention in the core developer community.

I would argue these are mostly cosmetic and adding clarity, the main content has been there since July, and it has been brought up almost every CC meeting since then with a request to review, [check the minutes](https://github.com/sktime/community-org/tree/main/community_council/previous_meetings).

I think you have also not made change requests as you say, at least I cannot find them.

aiwalter reviewed (changes requested):
Pending discussions at other PR

fkiraly commented:
@aiwalter, you reverted #3095 because you said you had change requests.

Please write explicitly what these change requests are, currently there are non that you have raised.

ltsaprounis reviewed (changes requested):
I believe that any issue / process around "resource allocation" has no place in the Code of Conduct of an OSS project. I had a look at the CoC of various projects within the scientific python community and didn't find anything similar. Here are some examples, in no particular order:
1. [pandas](https://pandas.pydata.org/community/coc.html)
2. [numpy](https://numpy.org/code-of-conduct/)
3. [scipy](https://docs.scipy.org/doc/scipy/dev/conduct/code_of_conduct.html)
4. [pytorch](https://github.com/pytorch/pytorch/blob/master/CODE_OF_CONDUCT.md)
5. [pymc](https://github.com/pymc-devs/pymc/blob/main/CODE_OF_CONDUCT.md)

Also, most CoC documents are concise and follow the same theme.
i.e. Here are our principles, here are the rules, here is how to report a problem and to who, here are the actions we will take as a project on CoC incidents.

ltsaprounis commented:
> That's argument from authority plus possibly cherry picking.

Apologies, I didn't take debate or philosophy class at school. 
So after googling [argument from authority](https://en.wikipedia.org/wiki/Argument_from_authority) I still can't see what's wrong with it. I can assure you I didn't do any cherry picking, if you still think that's the case, you can always do your own research and respond in this PR with another list of popular OSS projects that have a CoC similar to the one proposed here.

fkiraly commented:
@ltsaprounis, it would be great if you could talk to the matter instead of asserting that argument from authority is valid.

How about my question:

> For instance, systematically excluding people from the global South from internship or job opportunities is a problem, would you agree?

> Please answer yes/no.

> If you answer yes, which clause of the current CoC covers it currently (assuming the systematic exclusion from opportunities is accompanied by very polite and professional communication).

fkiraly commented:
> Is it possible to split up this new PR into smaller ones please?
Better yet un-merge the https://github.com/sktime/sktime/pull/3095 and just keep all the discussions and the code there?

@GuzalBulatova, #3095 is already unmerged, by @aiwalter.

It is not possible to unmerge a merged PR and make it open.

fkiraly commented:
> you can always do your own research and respond in this PR with another list of popular OSS projects that have a CoC similar to the one proposed here

@ltsaprounis, if we would have gone by the number of other OSS projects that do a certain thing as a metric that something is a good idea, we would not have started `sktime`, because at that point the list of scikit-learn like time series libraries (as far as I know) was length zero.

I'm not sure what else I should way about argument from authority and its validity.

fkiraly commented:
@ltsaprounis, since argument from authority seems to convince you.
First five hits on my google on how to write code of conduct how they appear in my search (excluding examples and including only guidelines on writing CoC)

1. https://www.indeed.com/hire/c/info/write-a-code-of-conduct

sections recommended

* Anti-harassment
* Financial integrity and compliance (reporting, payroll, business travel, purchasing)
* Data privacy
* Communications
* Client and business partner code of conduct
* Intellectual property and conflicts of interest

2. https://www.valamis.com/hub/code-of-conduct#what-to-include

content recommended:

* Company values - includes "business ethics", "social responsibility"
* Employee behaviour - includes "standards of professionalism" and "use of company assets"
* Internal Practices
* External Practices - includes "conflicts of interest"

3. https://www.business.qld.gov.au/running-business/employing/staff-satisfaction/staff-code-conduct/writing

* Ethical principles
* Values
* Accountability (incl "duty of care obligations" and "avoiding conflicts of interest")
* Standard of conduct
* Standard of practice
* Disciplinary actions

4. https://www.ethics.org/resources/free-toolkit/code-provisions/

* Compliance, Integrity and Anticorruption - includes financial reporting, expense records, bribes
* Conflicts of interest
* Employee, Client and Vendor Information
* Employment Practices - incl "proper use of authority", "equal opportunity", "discrimination"
* Environmental Issues
* Ethics and Compliance Resources
* Internet, social networking and social media
* Relationships with third parties

5. https://projectinclude.org/writing_cocs

long list, including:

(4) standards of professionalism
(5) use of company property
(13) Inappropriate use of company property or assets
(5)

fkiraly commented:
My summary for you, @ltsaprounis:

if you look at guidelines on writing CoC for entities that deal with and spend money, sections on resource allocation, decision making are common and considered mandatory.

Although the words used are often "company" related, e.g., "(mis)use of company assets", or accounting related, e.g., "financial integrity and compliance".

fkiraly commented:
@ltsaprounis, if I look at the publicly available CoC of your company, it also has similar sections.
(I'm not linking it here since I do not know whether you want that information to be publicly available)

I'm citing from it:

Conflicts of interest:
* We make sure our ability to perform our duties is not influenced by, or in conflict with, our personal relationships or any actual or anticipated benefit. A conflict of interest arises when personal activities, interests or relationships interfere or appear to interfere with our professional judgement, decisions or actions.
* We always keep in mind that potential conflicts can usually be resolved through open discussion and transparency

Financial integrity and transparency of reporting:
* The financial information that we maintain internally or provide to shareholders, regulators and other stakeholders, must be accurate, complete and timely.
* Do not delay or accelerate the recording of revenue or expenses between reporting periods.

Corruption:
* Bribery and corruption can include, but are not limitedto, cash payments, bank transfers, holidays disguised as business trips, facilitation payments, commissions, gifts, event tickets, job opportunities, donations, hospitality, discounts, memberships without a clear business purpose, political contributions, lobbying abuse, nepotism, undeclared conflicts of interest, misreporting of travel expenses, disproportional payments to a third party, or preferential terms.

fkiraly commented:
Does this satisfy your personal requirements on [arguments from authority](https://en.wikipedia.org/wiki/Argument_from_authority)?

fkiraly commented:
upon suggestion of @GuzalBulatova (slack), changed expectations on creating opportunities. In the new version, leaders are expected to create fair opportunities (without restriction), and members of the community are strongly encouraged (not expected ) to create fair opportunities as have been created for them.

fkiraly commented:
@aiwalter, I think I have actioned your two major points that you brought up:

* upon request of @aiwalter (slack), removed expectation that majority of benefits of an `sktime` commercial venture need to go to the community. What remains is the weaker requirement that it needs to go through transparent information channels and community decision making.
* upon request of @aiwalter (PR #3095), added clarification on the meaning of "resource"

Kindly re-review.

fkiraly commented:
@ltsaprounis, @Lovkush-A's post here contains two examples of well-funded open source projects (order of funding comparable to sktime) that have regulations on resource allocation, see here: https://github.com/sktime/sktime/pull/3095#issuecomment-1311834748

Your original point ("no open source project has regulations on resource allocation!") is therefore not valid? Kindly re-review and comment.

aiwalter commented:
I dont know even what an "sktime venture" should be? If 2 core devs have a liaison? If a PR is merged without review?😁 all sktime rights are defined in the license. Everything else is private  business of individuals and sktime has no right to claim anything. Anyone is free to use sktime to run a business or science project - thats the beauty of FOSS, so I dont get why you think you think you have anything to decide there that goes beyond whats in BSD-3 license?

fkiraly commented:
> so I dont get why you think sktime has anything to decide there that goes beyond whats in BSD-3 license?

@aiwalter, at minimum the BSD-3 license stipulates that we, the `sktime` project, can decide how our brand is used, that's quite a strong statement, here is the clause:

"Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission."

That means, yes, anyone is free to use `sktime` - but you cannot call your business "`sktime` consulting" or similar with an implied endorsement if you do not get it from the license holders.

So: no one, not me, not you, not anyone else, can go without express permission and do sth under the `sktime` brand, and we, as the community (represented through its bodies) can decide what the conditions are.

In addition, as a community, we can decide on how we handle corruption, bribery, conflict of interest and the like, e.g., whether we want to allow it under code of conduct or, maybe, ... not?

fkiraly commented:
@aiwalter, can you confirm you understand the distinction? Between using `sktime` as a piece of software, or using `sktime` as a brand, with possibly implied endorsement?

aiwalter commented:
Everything is written in license. I am able to read that

fkiraly commented:
@aiwalter, from my perspective, your answers remain elusive and unspecific to this PR.

Are you saying that, in general, we have a software license, so we do not need a code of conduct at all?

I don't see how your statements is specific, say, to the content of this PR, e.g., material on decision making, anti-corruption, conflicts of interest.

Are you perhaps saying that the BSD-3 license already covers well enough our community's expectations on how to spend money?

miraep8 commented:
I would be in favor of @Lovkush-A 's suggestion that we move the resource allocation stuff into a new PR, and potentially a different document?  

Personally do think that there are a lot of advantages of having a short, simple and concise CoC, it makes it easier for new people to read and for people to be fully aware of all the clauses.  With too much detail it runs the risk of becoming more of a legal tome!  I think that was what @ltsaprounis was getting at as well (perhaps? I don't mean to put words in anyone's mouth)  At the end of the day though it is a choice though, and I think we can come up with some successful examples of people doing it both ways, I think what should guide our choice is which option we collectively think we can make this a fairer, more inclusive and more accessible community.  My vote is for a short/concise document however! With flexibility to adapt to future circumstances. 🗽 

Given that sktime has been handling more funding/resources lately, and will probably continue to, I think it is worth stating some best principles about what resource allocation looks like _somewhere_ though and I do appreciate @fkiraly for starting the conversation.  Again, I think Lovkush's suggestion to start with a base set of principles and then build from there was a good one. 

Either way though I think the nature of this PR is also somewhat confusing (given the merge/unmerge/remerge timeline), so merging a simple version here - and then moving some of this over to a new PR could allow for a cleaner slate?

ltsaprounis commented:
Thanks @miraep8, yes I agree with you and @Lovkush-A on this. I never said that we shouldn't have a policy regarding the management of funds that are allocated to sktime, I just think that the CoC is not the right place for it. 

I agree with the idea of a separate PR for a new governance section that is dedicated on the above.

GuzalBulatova commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+Examples for in-scope:
+
+* an academic allocating grant monies from a grant, or a data science consultant providing consultancy services, with a primary ``sktime`` branding
+* hiring decisions for roles that are advertised with a primary ``sktime`` branding
```
First two examples are out-of-scope. Anyone can request money for their grant claiming they're going to post their research results in sktime. Anyone can open a business that sells forecasts and put the "we're widely using sktime", the way now people put "we use OS libraries like statsmodels and pandas".

fkiraly replied:
Not true: the BSD-3 license puts ``sktime`` branded activities within scope of license holder approval.

The key property is being ``sktime`` branded e.g., under a company name like "``sktime`` consulting", or a random university founding an ``sktime`` department.

Whereas a company named "``foo`` consulting", or a random university's "forecasting toolbox department" would not be impacted.

fkiraly replied:
the bullet points (due to the technical term "``sktime`` branded") apply to, and only to, activities that directly and prominently use the ``sktime`` name. They do not apply to similar activities that do not use the ``sktime`` name.

Is that subtlety clear?

To repeat: we have *legal* decision making power - supported almost worldwide by national and international legal systems! - on anything called ``sktime``, due to the BSD-3 license we are using.

Whereas the CoC has jurisdiction only where we can enforce it, i.e., among elected members.

GuzalBulatova replied:
"Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission." - I think that doesn't mean that people can't use sktime in the name of their project, etc, is that correct? Or did you mean some other part of the license that I missed?

fkiraly replied:
> I think that doesn't mean that people can't use sktime in the name of their project, etc, is that correct?

I think it does mean that.
No one can use ``sktime`` in their project or product, without permission, it does mean that.

This applies even without a trademark afaik, although the ease with which it can be legally enforced is impacted by whether it is registered.

GuzalBulatova commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+* an academic allocating grant monies from a grant, or a data science consultant providing consultancy services, with a primary ``sktime`` branding
+* hiring decisions for roles that are advertised with a primary ``sktime`` branding
+* spending decisions from ``sktime`` community administered accounts
+* decisions arising from academic collaboration or commercial requests directed to the ``sktime`` community,
```
n 3 - fully agree
n 4 - we can only demand this from community leaders (CC). We can't stop our contributors from agreeing to such collaboration if they saw a post in public channel. If our mentee sees academic opportunity posted like "looking for a researcher who has experience with sktime" they're very free to contact that person. If one of our contributors sees that there's a commercial request - same.

fkiraly replied:
> we can only demand this from community leaders

Sure, agreed on the community leader point. I clarified it in the scope statement that this is basically the scope of jurisdiction here, also including core developers and events (e.g., people intentionally destroying job ad posters around a "women in data science" special event could be tried for opportunity discrimination)

fkiraly replied:
> If our mentee sees academic opportunity posted like "looking for a researcher who has experience with sktime" they're very free to contact that person

Mentees in this scenario do not make hiring decisions, so they would not be impacted. They *get* hired (or not), so you would police the person who does the hiring, and only if they are part of `sktime` or do it on behalf of `sktime` (e.g., panelists on an `sktime` grant who are not elected `sktime` members but have delegated authority)

fkiraly replied:
Re consultancy, a mentee would be impacted if they say "I'm providing consultanty *on behalf of `sktime`*" and create the appearance of officiality.

If they put on their CV that they were an `sktime` mentee and say they have experience, but bill on a personal (not `sktime` branded company or activity), it's completely fine and they are not subject to the jurisdiction.

See the point on `sktime` branding above on the distinction.

GuzalBulatova replied:
Ok in my opinion a mentee does make a decision - to take the opportunity, and are free to do so. They can make a decision to advertise this opportunity further, right? If they take it themselves instead I see it as absolutely ok scenario

fkiraly replied:
of course, mentees and other community members can take on published job opportunities or requests for consultancy. They can also say they are contributors to `sktime`, on their CV. The line is at claiming they are "official `sktime` consultants" or the like (in a case where they are not). I would say "`sktime` developer" is already on the wrong side on that line, unless they are a core developer.

GuzalBulatova commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+e.g., university administrations,
+company policies, national anti-corruption and anti-bribery laws (or their absence), etc.
+
+For instance, it is in-principle possible to misappropriate ``sktime`` resource, while
```
Just clarifying, under "resource" we mean our community funds, correct? If yes, fully agree

fkiraly replied:
yes, and other things, as clarified in the new appendix section "what is meant by resource" that @aiwalter has requested.

GuzalBulatova replied:
Like said, I think that sktime has control only over the money that flows to sktime. Donations to OC, donations from industry, support from projects like GSoC and NumFOCUS and other forms of funds that come directly to the project.

fkiraly replied:
> I think that sktime has control only over the money that flows to sktime

Yes, I agree, in the sense that it also includes money that is *supposed to flow to sktime*, e.g., in the embezzlement scenario where someone redirects it maliciously.

GuzalBulatova replied:
Yes agree. That includes the money that was supposed to go to the sktime community, like donations. A research grant is not an instance of such, for example, because a grant goes to the research, not the OSS.

GuzalBulatova commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+Examples of not acceptable behaviour
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* obtaining a resource opportunity through ``sktime``,
```
on behalf of sktime - this is only applicable to CC members and in (very unlikely) situations like "My company would like to make a donation to sktime and I would like it to be through me." -> got the donation to their acct -> vanished with the donation.

fkiraly replied:
> this is only applicable to CC members

core devs too

GuzalBulatova commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+
+* obtaining a resource opportunity through ``sktime``,
+  then retreating or resigning ``sktime`` roles to declare that the opportunity is now in a personal role and not on behalf of ``sktime``.
+* pretending to communicate on behalf of ``sktime``, or unauthorized use of the ``sktime`` brand,
```
Agree that this is an inappropriate behaviour. However, the 'unauthorised' again goes against the open-source nature of the project. We do not authorise anyone to use the package - everyone is free to do so. With the brand and the logo - we should have (already have?) the TM to 'authorise' its use, otherwise we're not a company and can't stop anybody from using sktime as a brand if I'm not mistaken.

fkiraly replied:
I almost agree with you @GuzalBulatova, but I think you are not entirely correct about two points.

Legal point (best of my knowledge):

> we should have the TM to 'authorise' its use

We do not need the TM, clause 3 of the BSD-3 license suffices, and we can say "authorized or not" (see above)

Moral point:

> However, the 'unauthorised' again goes against the open-source nature of the project.

I disagree with this, open source does not mean anarchy or laissez faire.

That's why we have a code of conduct.

We have the legal right to control who says they are or represent ``sktime``, and morally we should not just give that away to anyone under any condition. E.g., someone who ticks all the possible boxes in violating our code of conduct.

fkiraly replied:
shorter summary:

* you are right, there is no authorization on use of software package and this would indeed contradict open source
* what this all refers to is *authorization on use of brand* (not software package). See discussion above.

GuzalBulatova commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+  then retreating or resigning ``sktime`` roles to declare that the opportunity is now in a personal role and not on behalf of ``sktime``.
+* pretending to communicate on behalf of ``sktime``, or unauthorized use of the ``sktime`` brand,
+  including but not limited to violations of the BSD 3-clause license
+* soft refusal to implement best anti-corruption and anti-bribery practice, e.g.,
```
I think this was already stated above that the resource allocation should be fair and comply with the anti-corruption practices? If this is intentional reiteration - np, can leave it here

GuzalBulatova commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+  including but not limited to violations of the BSD 3-clause license
+* soft refusal to implement best anti-corruption and anti-bribery practice, e.g.,
+  soft refusal to make budgets, invoives, or spending records transparent to the ``sktime`` community
+* attempts to suppress criticism of bad practice through invoking the code of conduct, e.g., tone policing or DARVO
```
This too was specified above, see line 177

fkiraly replied:
I think it is so common (conditional on the violation having occurred) that it is better to repeat.

GuzalBulatova replied:
Ok, I don't see this as a blocker, especially if it's intentional. 

However, I believe the longer the document the less likely that people are going to go through it. Our main intention here is that when people join they actually _read_ and _understand_ what the CoC implies and _agree_ to that. I think it's more likely if the CoC is written in plain and clear language: delivering clear and concise points on what's expected from the contributors. 

I do NOT claim that this is your intention, but the verboseness of CoC in my opinion contradicts the purpose a bit. That's why I'd avoid repetition and excessive elaboration. Maybe an option can be putting this in a separate document, sth like "a guide to the sktime CoCC"? That's intended for the CoCC members primarily and can be referred when making decisions/solving hard cases and contain examples, etc... Wdyt? We still should have all the points in the main CoC and only put the details, elaborations and examples in the CoCC guide.

fkiraly replied:
ok, I'm happy to splice the appendix on a separate page, if that's preferred?

GuzalBulatova replied:
I personally would prefer that, yes.

GuzalBulatova commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+
+DARVO, intimidation, noise generation, and plausible deniability may be common tactics found in an investigation.
+
+It is hence crucial to establish the facts.
```
I think it's crucial to establish the facts no matter the violation. I'd reduce this to this one line.
"The investigation's focus must be on the facts."

fkiraly replied:
ok, changed, have a look

GuzalBulatova commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+* hindering of the investigation itself, e.g., by not engaging with it to the fullest extent possible, creating community dissent around it, etc
+* promoting decision making on ``sktime`` governance that would weaken accountability,
+  transparency, decision making capacity, investigation practices, or adherence to best practice in resource allocation
+* despite better knowledge - evidence that the decision maker is aware of best practice and nonetheless acts against it.
```
I think "Ignorance of the law is no excuse" should be here. If a CC member claims that they simply didn't know they couldn't transfer the OC money to their own pocket - it's not an excuse, no `evidence that the decision maker is aware of best practice` is needed.

fkiraly replied:
excellent point - I have added this as clarification. Let me know if you would suggest to improve the formulation.

GuzalBulatova reviewed (changes requested):
Agree with many points, disagree with some. 

The main blocker for me is `A.2  Guidance on applying the CoC for resource allocation decisions`, especially the Scope section. Some points there are inapplicable to an open-source project like sktime. We put the code out for everybody to use, whatever they use it for is outside the scope of the project, we're not responsible for that. People are free to benefit from sktime, even without a single contribution for that matter! Just like millions of people profiting from sklearn, statsmodels, pandas, numpy etc. People make whole businesses based on tuning OSSs like postgreSQL and profit from it - that's one way to use OSS.

As long as the "resource" in the "resource allocation" are sktime's own funds - what we were donated (to Open Collective, got support from NumFOCUS), I fully agree that we should handle them fairly. That is without favouring/discriminating the beneficiaries of the resource allocation, and we should prioritise the community benefit, etc. 

What people make from using our package - well that's what is absolutely out of scope for sktime. Our contributors are aware that they're submitting their work out for everybody to use and don't expect a ROI. Their profit is that they're (hopefully) having fun/they've learn sth/they need that piece of code for their job or research project/any other reason these amazing people are dedicating their time to contribute.

fkiraly commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -9,14 +9,15 @@ experience. Accordingly, everyone who participates in the sktime project
 
-We are particularly motivated to support new and/or anxious
+We are dedicated to a **fair and equal opportunities environment** for everyone,
+and therefore are particularly motivated to support new and/or anxious
```
I have not added the "anxious", this is preexisting. Agree that it should be rephrased though, a bit odd.

fkiraly commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -104,6 +129,11 @@ hope you will consider to be appropriate community guidelines:
    succinct so as to allow maximum engagement by all participants. Do
    not interrupt others on the basis of disagreement; hold such comments
    until they have finished speaking.
+-  **Do not hesitate to voice critical opinions, and accept constructive
```
> We can't enforce accepting the constructive criticism

We can outline our expectations.

The python CoC has "gracefully accept constructive criticism", and I think that should be an expectation.

Of course we can enforce that people should accept constructive criticism.

Failure mode: 

"A: I think this code can be improved by changing X to Z."

"B: You have deeply upset me by implying my code needs improvement, or that I have been unable to see that my code needs improvement. This is a clear violation of our code of conduct, and I want you to refrain from hurtful behaviour like this in the future. Please also work on your communication skills."

GuzalBulatova replied:
This scenario is exactly what I meant by "fight", moreover, it's a false accusation of CoC violation.
What I mean by "not enforcing" is personal borders, the person doesn't have to change their personal opinion on the matter as long as expression of this opinion doesn't violate CoC. Their PR just not going to be approved and merged - if that's what you mean under "enforce" then I agree.

fkiraly replied:
> What I mean by "not enforcing" is personal borders, the person doesn't have to change their personal opinion on the matter as long as expression of this opinion doesn't violate CoC

Yes, I see what you mean. I think what I mean under "enforce" is what you understand, yes.

In common parlance (as far I understand it), "accepting criticism" means acknowledging it without a fight, resistance, or retaliation like above.

It does not mean the recipient has to change their opinion, that would be absurd, because any position can be criticized.

fkiraly commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -133,13 +163,21 @@ project event or platform include:
+-  decision making on resource allocation which has the effect of excluding people
+   on the basis of protected characteristics, or other forms of systemic discrimination
+-  decision making on resource allocation that confers advantages to personal networks,
+   and excludes others, e.g., preventing advertising of opportunities, not applying best practice in recruitment
```
ok

fkiraly replied:
done

fkiraly commented:
@miraep8, to answer your points:

> I would be in favor of @Lovkush-A 's suggestion that we move the resource allocation stuff into a new PR, and potentially a different document?

My concern is that this may increase confusion and make the PR more vulnerable to stalling tactics.

But I'm happy to do that if @aiwalter, @GuzalBulatova and @ltsaprounis tell me they would be fine with this and promise to give it speedy (possibly conditional) approval.

> Personally do think that there are a lot of advantages of having a short, simple and concise CoC

Agreed!

Although kindly note the history. Originally, what was added were just a few bullet points and paragraphs, i.e., excluding appendices.

Then this got a lot of "but you need to explain more clearly what this means", e.g., "explain how it is applied, I'm concerned this is applied arbitrarily", "what does resource mean" etc.

These requests increased the size of the PR by what is in the appendix that contains these clarifications.

It is in parts some of the same people who say this should be shorter, who back then said, this needs much more details.

This does not feel right to me.

fkiraly commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -133,13 +163,21 @@ project event or platform include:
+-  decision making on resource allocation that confers advantages to personal networks,
+   and excludes others, e.g., preventing advertising of opportunities, not applying best practice in recruitment
+-  reducing transparency of operations of sktime, e.g., by obfuscation of the paper trail
+-  invoking the code of conduct dishonestly, or as a form of bullying,
```
disagreed, ruling should be on the balance of probability, based in turn on circumstantial evidence.

There should be burden on the accused to cooperate with proceedings, e.g., to provide proof for how they have handled money in a case of invoice fraud or similar.

Otherwise, anyone can commit a crime and then feel safe by hiding the evidence.

GuzalBulatova replied:
I see what you mean, but in line with "assume best intentions" our basic assumption should be that there's been no violation until proven so, and that there should be evidence to prove the wrongdoing proportional to the accusation. Agree that the person who's accused of sktime funds misappropriation for example should have documentation on these funds handling and provide it. 
But we can't just have people going around accusing other community members of any CoC violation they can think about without any facts to back it up. If it's just "my words against your words" then we look at the circumstantial evidence, like what are parties' motivations, history, etc, agree of course, but if there's no substantial proof (circumstantial evidence and motivation counts as proof) - then there's a false accusation, which is also inappropriate and a form of wrongdoing. I thought that's partially what's meant under the "invoking the CoC dishonestly"?

fkiraly replied:
>  our basic assumption should be that there's been no violation until proven so, and that there should be evidence to prove the wrongdoing proportional to the accusation. 

fully agreed - all I'm saying, the evidence should be "on the balance of probability", which is a lower standard but common in investigations, say, of research misconduct where the investigating party does not have the rights a judge or the police might have.

See as an example the [UCL guidelines to investigate research misconduct](https://www.ucl.ac.uk/governance-compliance/research-governance), this and other similar documents put the standard of evidence as "on the balance of probabilities".

> there should be evidence to prove the wrongdoing proportional to the accusation. 

That's what it basically means.

> but if there's no substantial proof (circumstantial evidence and motivation counts as proof) - then there's a false accusation, which is also inappropriate and a form of wrongdoing.

The typical standard that justifies making an accusation is typically that it the suspicion is plausible, given the information that the accuser has. It requires no collection of evidence on the side of the accuser.

E.g., to put someone under the suspicion of financial misconduct it suffices if, say, balance sheets are not published in line with institutional expectations. Whereas, for a conviction there must be evidence that the balance sheets were doctored - but the investigation panel has the powers to obtain the balance sheets, or to request them.

> I thought that's partially what's meant under the "invoking the CoC dishonestly"?

Yes, correct. The standard for proving that the CoC has been invoked dishonestly is the same, on the balance of probabilities.

fkiraly replied:
To answer the implicit point you are making, I don't think the proposed changes would allow accusation or investigation based on hearsay or made up grounds.

Because the standard of evidence is explicitly set to "on the balance of probabilities".

fkiraly replied:
Regarding the burden of proof, I don't think it should lie with the accuser.

I think the process should be such: the CoC committee determines whether the accusation is plausible, then possible immediate actions. If the CoC committee thinks the accusation is plausible, they move to investigate.

It should not be the responsibility of the accuser to investigate or produce proof.

For instance, let's assume a participant of an `sktime` event says they were sexually harassed at an `sktime` event. Things like interviewing witnesses, getting security camera recordings etc are not the burden of the person that was harassed.

It would be similar if the police gets involved - it is the police that investigates first, not the person that was harassed.

fkiraly replied:
Same point in a bullying context, made coincidentally by a recent article in HBR:
https://hbr.org/2022/11/how-bullying-manifests-at-work-and-how-to-stop-it?utm_source=pocket-newtab-global-en-GB

Citing:

> Placing the burden of proof and anti-bullying work on the target ignores the fact that bullying is trauma, and that for most people, documenting their own trauma while it is happening while maintaining productivity is an impossible demand

fkiraly replied:
(also makes very similar points about resource allocation)

GuzalBulatova replied:
What I am saying is that CoC shouldn't accept "bare words" as enough to condemn the person who was accused. Z can say that X is harassing them and not provide any proof because "well you know that was very traumatic event for me", hence the CoC has to simply accept that Z is telling the truth? 
Compare this with Z showing messages / providing a reason why they don't have sufficient proof like "we were at B location on the conference where there were nobody else but me and X, X behaved in ABC manner and told me DEF, and I have shared my concerns with my friend W right after the incident, you can talk to them". 
If there's something serious then the police should be involved and we should wait for the results of the investigation.
I stand by the statement that "It is better that ten guilty persons escape than that one innocent suffer".

fkiraly commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -168,13 +206,13 @@ GitHub.
-In case of conflict of interest of sktime CoC committee members, you can report
-through the [NumFOCUS CoC process](https://numfocus.org/code-of-conduct),
+In case of conflict of interest of sktime CoC Committee members, you can report
+through the `NumFOCUS CoC process <https://numfocus.org/code-of-conduct>`__,
```
yes, agree in principle.

This comment is however entirely out of scope for this PR, we are not deciding on composition of CoC.

CC should not be an alternate contact point due to conflict of interest and separation of powers.

GuzalBulatova replied:
I've checked CoCs of other projects under Numfocus and found a reference to escalation to their CoCC, so I'm fine with it.

fkiraly commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -207,7 +245,7 @@ Dr Franz Király (f.kiraly@ucl.ac.uk)
 We encourage community members to step up to become a member of the committee.
 
-You can use the NumFOCUS CoC reporting process as an alternative entry point
+You can use the `NumFOCUS CoC process <https://numfocus.org/code-of-conduct>`__ as an alternative entry point
```
disagreed, that would discourage people who need help to seek help.

GuzalBulatova replied:
Posted above, agree, resolve

fkiraly commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+Guidance rules:
+
+* the less powerful party should be protected, i.e., CoC cannot be invoked by lead community member to silence critique of how funds are spent.
+  But, CoC should protect new contributors or early career members from more powerful community members, e.g., in operational or technical disputes.
```
it means:

* they should try convince the core developers, say, by producing evidence that the process was fair. The burden of proof should be with leadership, to show the process was fair.
* the CoC will not move to silence the early career member in any way, as long as the allegations are plausible and civil in tone.
* in particular, leadership just has to stomach the accusations as long as they remain even vaguely plausible. It's a moment of accountability.

GuzalBulatova replied:
Strongly disagree with "vaguely plausible". By proof I don't mean that the person suspecting favouritism has to provide the videos/messages of this leader discussing how they're going to hire their favourite. Just stating the facts that, for example, this leader had candidate A that has the same level of competence and is ready to start the project at the same time as B but chooses B, and B happens to be his nephew - is a proof and it must be provided by the person complaining to the CoCC. This person can not accuse leadership of favouritism just based on their impressions or feelings, zero facts.

fkiraly replied:
> This person can not accuse leadership of favouritism just based on their impressions or feelings, zero facts.

I think the standard should be, "plausible suspicion".
The plausible suspicion must be outlined to the CoC committee for an investigation.

If the community member decides to make it public in the community based on plausible suspicion, I think that should be ok, in the interest of accountability, checks and balances.

Similar to how it is fine for free media to allege wrongdoing of public role holders on a basis of plausibility - in countries that allow freedom of speech, the rules are less stringent in protecting public persons because of the public interest and the accountability moment.

fkiraly replied:
In particular, I would *strongly disagree* that the community member must be silent until the CoC committee has completed their investigation, say.

They only need to be civil in tone, but still can present the evidence basis for their suspicion.

This should *not* be stifled or silenced by invocation of things like "individual discomfort" (of the person in power).

fkiraly commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,161 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+
+In scope for this section are all resource allocation decisions that satisfy at least one of the following criteria:
+
+* any decisions that allocate resources primarily obtained due to the ``sktime`` project and the volunteer contributions of the community.
```
I see what you mean and how it could be construed to be in conflict with the BSD-3 license.

It is meant to apply to elected members only, i.e., core developers. It is not 100% clear from the text, so I will clarify that.

I.e., the gist should be, you are elected to represent `sktime` (as core dev, CC member etc), so you need to act on behalf of the community rather than on your own behalf.

If you are not an elected member, do as you wish as long as you comply with the BSD-3 license.

fkiraly replied:
I have clarified the scope of jurisdiction.

fkiraly replied:
see above - I think it should now be 100% clear that the software distribution license is not impacted, that this applies primarily to elected members of communities in executing their roles (or operating in grey spaces between their sktime role and other roles)

GuzalBulatova replied:
I don't think core-devs bear such burden, do they have access to sktime funds / OC? I thought that's the responsibility that only CC has? Acting on behalf of the community, etc.

GuzalBulatova replied:
" obtained due to the ``sktime`` project" is hard to define. For example, if a person gets hired on a position and one of the points of their resume are "a core-dev of sktime" - do they now obtain their salary due to sktime project and have to give up some of it? Maybe a start-up project is named "sktime product demand forecasting" and gets investment, how to define if the investors made their decision primarily due to "sktime" in the name and not the "product demand forecasting" part?

fkiraly replied:
> I don't think core-devs bear such burden, do they have access to sktime funds / OC? I thought that's the responsibility that only CC has? Acting on behalf of the community, etc.

Core devs have a lesser version of rights of this kind, but they still do have them, e.g., when organizing an event, handling funds, handling social media, etc, on behalf of the community or in delegation.

Formally, if you look at our main gov document, *core developers participate in decision making*.

So, of course, rules on decision making should include them in theie key scopes, in my opinion.

fkiraly commented:
@miraep8, @Lovkush-A, @ltsaprounis - if we were to split, how/where would you suggest to split?

fkiraly commented:
> Thanks @miraep8, yes I agree with you and @Lovkush-A on this. I never said that we shouldn't have a policy regarding the management of funds that are allocated to sktime, I just think that the CoC is not the right place for it.

@ltsaprounis, would the location (in isolation) be a blocker for you?

Would changing location or composition of the additions unblock you? If so kindly provide details and I will be happy to accommodate.

Re what @Lovkush-A said, if you read this carefully: https://github.com/sktime/sktime/pull/3095#issuecomment-1311834748
I would interpret what he said as: CoC is as good a place as any, better have this in a random place where it is being read than not have it (subject maybe to minor changes on points worthy of discussion).

fkiraly commented:
@aiwalter, I was originally unable to parse your feedback, apologies.

With @GuzalBulatova's, I think you may have been making the same points on the BSD-3 license? I.e., you were concerned we would override it and put a burden on every user?

If so, then my clarifications (that this is not the case) towards @GuzalBulatova's remark should address your main point of criticism as well? Would appreciate if you could have another look.

Lovkush-A commented:
> My concern is that this may increase confusion and make the PR more vulnerable to stalling tactics.

I dont think it should be confusing. And I think it is better to get the things we agree on in, and only have the 'controversial' bits stalled, rather than the whole thing stalled.

> if we were to split, how/where would you suggest to split?

I have not looked at this in a few weeks, so my memory is hazy. But I would suggest something like:
- have all the items not related to resource allocation in one pr (i remember chris saying he agreed with basically all of the non resource allocation items)
- have a 'minimally controversial' start on resource allocation items (likely in separate document?).
  - so start with things that should reach consensus quickly, e.g. funds in sktime open collective.
- over time, get scope increased to deal with more of the grey areas / hard to define things. i suspect this will take long and will be dragged out, but i think thats just how these things go.

fkiraly commented:
> I have not looked at this in a few weeks, so my memory is hazy. But I would suggest something like:
> have all the items not related to resource allocation in one pr (i remember chris saying he agreed with basically all of the non resource allocation items)

Ok, I'm not sure whether this makes sense, @Lovkush-A .
Resource allocation decisions are the main topic of this PR. Would that not mean remove everything?

Here's what @chrisholder said, in https://github.com/sktime/sktime/pull/3095#issuecomment-1304076136:

> I am fine with the concepts this pr puts forward for resolving conflicts but I disagree with 90% of the 'resource allocation decision' section because it's essentially a means of exerting claim to resources that Sktime doesn't have a right to claim.

I was confused by that, because the PR does not contain content on conflict resolution?

It does contain a section about how to resolve trade-offs which could have been meant, but this heavily make reference to resource allocation content. E.g., how to trade off leadership's rights with requirements on being accountable and answering questions about resource allocations.

As said, I would appreciate, if you or @chrisholder would concretely said what should go where in a split. Optimally, section numbers, line numbers, etc, rather than ambiguous references.

GuzalBulatova commented:
> Although kindly note the history. Originally, what was added were just a few bullet points and paragraphs, i.e., excluding appendices.
> 
> Then this got a lot of "but you need to explain more clearly what this means", e.g., "explain how it is applied, I'm concerned this is applied arbitrarily", "what does resource mean" etc.
> 
> These requests increased the size of the PR by what is in the appendix that contains these clarifications.
> 
> It is in parts some of the same people who _now_ say this should be shorter, who _back then_ said, this needs much more details.
> 
> This does not feel right to me.

@fkiraly I have checked the history, as suggested above. Please see comparison between the first commit and the latest one. There's a +46 lines difference, but to me it doesn't look like the original 225 lines can be described as minor additions.
<img width="1405" alt="Screenshot 2022-12-03 at 18 31 33" src="https://user-images.githubusercontent.com/73598322/205456475-aad1b84e-47f3-4a68-84ad-83d98d334b9a.png">
<img width="1405" alt="Screenshot 2022-12-03 at 18 31 55" src="https://user-images.githubusercontent.com/73598322/205456485-a5c994fe-1528-40d5-a61f-4c6a54b3ef37.png">

GuzalBulatova commented on docs/source/get_involved/code_of_conduct.rst:
```diff
@@ -342,14 +397,183 @@ Guide <https://www.djangoproject.com/conduct/>`__ and we are also
+* budget, money, financial credit
+* job opportunities, business opportunities
+* decision making power, soft or hard influence
+* dedicated work time of subordinates or dependents
```
How are we going to control that, I don't understand. Is this implying that we're affecting how someone's distributing their subordinates time and if it's "supposed to" go on working on sktime we're somehow able to influence that? How is that sktime's business and not the employer of said individual whose subordinates are mentioned here?

GuzalBulatova replied:
And re job opportunities - this is relevant imo only if the job opportunity was to sktime community. If someone messages a CC member with "Hi X, I see your contributions to sktime, I'd like to offer you a job" - it's not a "sktime resourse"

GuzalBulatova reviewed (changes requested):
I still disagree with some parts on resource allocation.

fkiraly commented:
Summary of most recent changes:

* this PR has now been open since July, here is the original: https://github.com/sktime/sktime/pull/3095
* it has experienced some serious engagement by contributors
* but also what looks like a lot of stalling tactics - e.g., people attempting to block the PR, and then not reacting, despite it being brought up multiple times at CC meetings
* combined with the fact that [the same people are involved in a recent attempt to capture `sktime` governance mechanisms and mobbing out of contributors](https://github.com/sktime/sktime/issues/3865), I am making the decision to disregard these blocks and merge.
* overall, by the current document - which clearly refers to me as the project PI and last authority - I have the power to do this, as it refers to me in person.
* An earlier version of this document would have changed accountability from me as a person to the full CoC committee, but for now I have left the reference to my person as unchanged. Given current code of conduct relevant matters and attempted resource capture, this is as a safeguard.
* as a minor relevant item, the earlier code of conduct reporting email has been captured by Martin Walter, and I have been locked out. I have changed this hence to a community email where I still ahve access.
